### ISSUE TITLE
Jenkins plugin design

### ISSUE THREAD

**Issue by narrowizard** [merico-dev/lake#307]
## Description
We need get metircs in #259 from jenkins api. columns we need is as following:
- build_name
- build_start_time
- build_end_time
- build_time
- is_build_success
- others...

## Describe the solution you'd like
- using a go jenkins sdk for api requests. (https://github.com/bndr/gojenkins)

### steps
1. get all job names - https://pkg.go.dev/github.com/bndr/gojenkins#Jenkins.GetAllJobNames
2. get all build ids for each job - https://pkg.go.dev/github.com/bndr/gojenkins#Job.GetAllBuildIds
3. get build detail for each build - https://pkg.go.dev/github.com/bndr/gojenkins#Job.GetBuild
4. since the response data is too long, I've added it in comments below.

## Entities design
```go
// TODO
type JenkinsJob struct {
	ID    int
	Name  string
	Class string
	Color string
}

type JenkinsBuild struct {
	JobID             uint // refer to JenkinsJob.ID
	Duration          float64 // build time
	DisplayName       string  // "#7"
	EstimatedDuration float64
	Number            int64
	Result            string
	Timestamp         int64 // start time
}
```

## Constraint
- we need jenkins data from past half year.

**Comment by narrowizard** [comment]
## jobs data
```json
{
  "_class": "org.jenkinsci.plugins.workflow.job.WorkflowJob",
  "Actions": [
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    }
  ],
  "buildable": true,
  "Builds": [
    {
      "Number": 66,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/66/"
    },
    {
      "Number": 65,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/65/"
    },
    {
      "Number": 64,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/64/"
    },
    {
      "Number": 63,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/63/"
    },
    {
      "Number": 62,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/62/"
    },
    {
      "Number": 61,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/61/"
    },
    {
      "Number": 60,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/60/"
    },
    {
      "Number": 59,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/59/"
    },
    {
      "Number": 58,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/58/"
    },
    {
      "Number": 57,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/57/"
    },
    {
      "Number": 56,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/56/"
    },
    {
      "Number": 55,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/55/"
    },
    {
      "Number": 54,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/54/"
    },
    {
      "Number": 53,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/53/"
    },
    {
      "Number": 52,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/52/"
    },
    {
      "Number": 51,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/51/"
    },
    {
      "Number": 50,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/50/"
    },
    {
      "Number": 49,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/49/"
    },
    {
      "Number": 48,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/48/"
    },
    {
      "Number": 47,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/47/"
    },
    {
      "Number": 46,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/46/"
    },
    {
      "Number": 45,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/45/"
    },
    {
      "Number": 44,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/44/"
    },
    {
      "Number": 43,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/43/"
    },
    {
      "Number": 42,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/42/"
    },
    {
      "Number": 41,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/41/"
    },
    {
      "Number": 40,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/40/"
    },
    {
      "Number": 39,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/39/"
    },
    {
      "Number": 38,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/38/"
    },
    {
      "Number": 37,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/37/"
    },
    {
      "Number": 36,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/36/"
    },
    {
      "Number": 35,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/35/"
    },
    {
      "Number": 34,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/34/"
    },
    {
      "Number": 33,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/33/"
    },
    {
      "Number": 32,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/32/"
    },
    {
      "Number": 31,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/31/"
    },
    {
      "Number": 30,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/30/"
    },
    {
      "Number": 29,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/29/"
    },
    {
      "Number": 28,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/28/"
    },
    {
      "Number": 27,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/27/"
    },
    {
      "Number": 26,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/26/"
    },
    {
      "Number": 25,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/25/"
    },
    {
      "Number": 24,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/24/"
    },
    {
      "Number": 23,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/23/"
    },
    {
      "Number": 22,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/22/"
    },
    {
      "Number": 21,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/21/"
    },
    {
      "Number": 20,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/20/"
    },
    {
      "Number": 19,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/19/"
    },
    {
      "Number": 18,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/18/"
    },
    {
      "Number": 17,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/17/"
    },
    {
      "Number": 16,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/16/"
    },
    {
      "Number": 15,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/15/"
    },
    {
      "Number": 14,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/14/"
    },
    {
      "Number": 13,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/13/"
    },
    {
      "Number": 12,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/12/"
    },
    {
      "Number": 11,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/11/"
    },
    {
      "Number": 10,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/10/"
    },
    {
      "Number": 9,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/9/"
    },
    {
      "Number": 8,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/8/"
    },
    {
      "Number": 7,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/7/"
    },
    {
      "Number": 6,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/6/"
    },
    {
      "Number": 5,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/5/"
    },
    {
      "Number": 4,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/4/"
    },
    {
      "Number": 3,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/3/"
    },
    {
      "Number": 2,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/2/"
    },
    {
      "Number": 1,
      "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/1/"
    }
  ],
  "color": "aborted",
  "concurrentBuild": false,
  "description": "Trigger all  rules check immediately.",
  "displayName": "Auto Init Gitlab",
  "displayNameOrNull": null,
  "downstreamProjects": null,
  "FirstBuild": {
    "Number": 1,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/1/"
  },
  "fullName": "Auto Init Gitlab",
  "fullDisplayName": "Auto Init Gitlab",
  "healthReport": [
    {
      "description": "Build stability: No recent builds failed.",
      "iconClassName": "icon-health-80plus",
      "iconUrl": "health-80plus.png",
      "score": 100
    }
  ],
  "inQueue": false,
  "keepDependencies": false,
  "lastBuild": {
    "Number": 66,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/66/"
  },
  "lastCompletedBuild": {
    "Number": 66,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/66/"
  },
  "lastFailedBuild": {
    "Number": 44,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/44/"
  },
  "lastStableBuild": {
    "Number": 21,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/21/"
  },
  "lastSuccessfulBuild": {
    "Number": 21,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/21/"
  },
  "lastUnstableBuild": {
    "Number": 0,
    "URL": ""
  },
  "lastUnsuccessfulBuild": {
    "Number": 66,
    "URL": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/66/"
  },
  "name": "Auto Init Gitlab",
  "nextBuildNumber": 67,
  "property": [
    {
      "parameterDefinitions": null
    },
    {
      "parameterDefinitions": null
    },
    {
      "parameterDefinitions": null
    },
    {
      "parameterDefinitions": [
        {
          "defaultParameterValue": {
            "name": "gitlabHost",
            "value": "http://i.test.meri.co:8022/"
          },
          "description": "",
          "name": "gitlabHost",
          "type": "StringParameterDefinition"
        },
        {
          "defaultParameterValue": {
            "name": "accessToken",
            "value": null
          },
          "description": "",
          "name": "accessToken",
          "type": "PasswordParameterDefinition"
        },
        {
          "defaultParameterValue": {
            "name": "userNumber",
            "value": "20"
          },
          "description": "最少要1",
          "name": "userNumber",
          "type": "StringParameterDefinition"
        },
        {
          "defaultParameterValue": {
            "name": "projectNumber",
            "value": "200"
          },
          "description": "最少要1",
          "name": "projectNumber",
          "type": "StringParameterDefinition"
        }
      ]
    }
  ],
  "queueItem": null,
  "scm": {},
  "upstreamProjects": null,
  "url": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/",
  "jobs": null,
  "primaryView": null,
  "views": null
}
```
## build data
```json
{
  "Actions": [
    {
      "parameters": [
        {
          "Name": "gitlabHost",
          "Value": "http://i.test.meri.co:8022/"
        },
        {
          "Name": "accessToken",
          "Value": ""
        },
        {
          "Name": "userNumber",
          "Value": "2"
        },
        {
          "Name": "projectNumber",
          "Value": "1000"
        }
      ],
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": [
        {
          "_class": "hudson.model.Cause$UserIdCause",
          "shortDescription": "Started by user Merico",
          "userId": "merico",
          "userName": "Merico"
        }
      ],
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": [
        {
          "_class": "jenkins.model.CauseOfInterruption$UserInterruption"
        }
      ],
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": [
        {
          "_class": "jenkins.model.CauseOfInterruption$UserInterruption"
        }
      ],
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": [
        {
          "_class": "jenkins.model.CauseOfInterruption$UserInterruption"
        }
      ],
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    },
    {
      "parameters": null,
      "causes": null,
      "buildsByBranchName": null,
      "lastBuiltRevision": {
        "SHA1": "",
        "branch": null
      },
      "remoteUrls": null,
      "scmName": "",
      "mercurialNodeName": "",
      "mercurialRevisionNumber": "",
      "subdir": null,
      "TotalCount": 0,
      "UrlName": ""
    }
  ],
  "artifacts": [],
  "building": false,
  "builtOn": "",
  "changeSet": {
    "items": null,
    "kind": "",
    "revision": null
  },
  "changeSets": [],
  "culprits": [],
  "description": null,
  "duration": 32645109,
  "estimatedDuration": 47511,
  "executor": null,
  "displayName": "#66",
  "fullDisplayName": "Auto Init Gitlab #66",
  "id": "66",
  "keepLog": false,
  "number": 66,
  "queueId": 14755,
  "result": "ABORTED",
  "timestamp": 1615304450843,
  "url": "https://kratos.meri.dev/job/Auto%20Init%20Gitlab/66/",
  "mavenArtifacts": null,
  "mavenVersionUsed": "",
  "FingerPrint": null,
  "runs": null
}
```

**Comment by narrowizard** [comment]
Closed by #312
